src/Proto.Remote.GrpcCore/GrpcCoreRemote.cs

```diff
@@ -0,0 +1,113 @@
+//   </copyright>
+// -----------------------------------------------------------------------
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Grpc.Health.V1;
+using Grpc.HealthCheck;
+using JetBrains.Annotations;
+using Microsoft.Extensions.Logging;
+
+namespace Proto.Remote.GrpcCore
+{
+    [PublicAPI]
+    public class GrpcCoreRemote : IRemote
+    {
+        private static readonly ILogger Logger = Log.CreateLogger<GrpcCoreRemote>();
+        private EndpointManager _endpointManager = null!;
+        private EndpointReader _endpointReader = null!;
+        private HealthServiceImpl _healthCheck = null!;
+        private Server _server = null!;
+        private readonly GrpcCoreRemoteConfig _config;
+
+        public GrpcCoreRemote(ActorSystem system, GrpcCoreRemoteConfig config)
+        {
+            System = system;
+            _config = config;
+            System.Extensions.Register(this);
+        }
+        public bool Started { get; private set; }
+        public ActorSystem System { get; }
+        public RemoteConfigBase Config => _config;
+
+        public Task StartAsync()
+        {
+            lock (this)
+            {
+                if (Started)
+                    return Task.CompletedTask;
+                var channelProvider = new GrpcCoreChannelProvider(_config);
+                _endpointManager = new EndpointManager(System, Config, channelProvider);
+                _endpointReader = new EndpointReader(System, _endpointManager, Config.Serialization);
+                _healthCheck = new HealthServiceImpl();
+                _server = new Server
+                {
+                    Services =
+                    {
+                        Remoting.BindService(_endpointReader),
+                        Health.BindService(_healthCheck)
+                    },
+                    Ports = { new ServerPort(Config.Host, Config.Port, _config.ServerCredentials) }
+                };
+                _server.Start();
+
+                var boundPort = _server.Ports.Single().BoundPort;
+                System.SetAddress(Config.AdvertisedHost ?? Config.Host, Config.AdvertisedPort ?? boundPort
+                );
+                _endpointManager.Start();
+
+                Logger.LogDebug("Starting Proto.Actor server on {Host}:{Port} ({Address})", Config.Host, boundPort,
+                    System.Address
+                );
+                Started = true;
+                return Task.CompletedTask;
+            }
+        }
+
+        public Task ShutdownAsync(bool graceful = true)
+        {
+            lock (this)
+            {
+                if (!Started)
+                    return Task.CompletedTask;
+                try
+                {
+                    if (graceful)
+                    {
+                        _endpointManager.Stop();
+                        _server.ShutdownAsync().GetAwaiter().GetResult();

```

Comment:
This already returns a task, why not use await?

Reply:
We can't await in a lock block.

Reply:
You could go with a semaphore then. But if we never re-use the remote after shutdown, you could just set a flag that it is already shut down in a lock..
The rest could be done after the lock block?